// msmc/dist/mjs/util/tokens.js
import { Minecraft } from "../auth/minecraft.js";
import { validate } from "./validate.js";
export function fromToken(auth, token, refresh) {
    if (validate(token) && refresh)
        return new Promise(async (done, reject) => {
            const xbl = await auth.refresh(token.refresh);
            try {
                done(await xbl.getMinecraft());
            }
            catch (e) {
                reject(e);
            }
        });
    let mc = new Minecraft(token.mcToken, token.profile, auth, token.refresh, token.exp);
    return mc;
}
/**
 * Gets a Minecraft token from a saved mcToken.
 * @param auth A new instance of the Auth object
 * @param token The mcToken
 * @returns A newly serialized Minecraft Token.
 *
 * @warning The Xbox object may not be restored using this method!
 */
export function fromMclcToken(auth, token, refresh) {
    return fromToken(auth, {
        mcToken: token.access_token,
        refresh: token.meta?.refresh,
        exp: token.meta?.exp,
        profile: { id: token.uuid, name: token.name },
        xuid: token.meta?.xuid,
    }, refresh);
}
