// msmc/dist/mjs/auth/social.js
import fetch from "node-fetch";
import { XPlayer } from "../util/xPlayer.js";
export class Social {
    auth;
    constructor(auth) {
        this.auth = auth;
    }
    async getProfile(xuid) {
        const profile = await this.xGet("/profile/settings?settings=GameDisplayName,GameDisplayPicRaw,Gamerscore,Gamertag", xuid);
        return new XPlayer(profile.profileUsers[0], this);
    }
    async getFriends(xuid) {
        const friends = await this.xGet("/profile/settings/people/people?settings=GameDisplayName,GameDisplayPicRaw,Gamerscore,Gamertag", xuid);
        let R = [];
        friends.profileUsers.forEach((element) => {
            R.push(new XPlayer(element, this));
        });
        return R;
    }
    async xGet(endpoint, xuid) {
        const target = xuid ? `xuid(${xuid})` : "me";
        let profileRaw = await fetch(`https://profile.xboxlive.com/users/${target}/${endpoint}`, {
            headers: {
                "Content-Type": "application/json",
                "x-xbl-contract-version": "2",
                Authorization: this.auth,
            },
        });
        return await profileRaw.json();
    }
}
export default Social;
